### PR TITLE
Add `-Ydebug-tree-with-id n`

### PR DESCRIPTION
**Pull request by nicolasstucki**
This postion allow descovering the place creation of a tree given its unique id.
Adding is as a setting to simplify debuging on a released version of dotty.

### CODE REVIEW

**Comment by nicolasstucki** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/ast/Positioned.scala [hunk 1]
```diff
@@ -23,11 +23,14 @@ abstract class Positioned(implicit @constructorOnly src: SourceFile) extends Pro
...
-    // assert(id != 2523, this)
-    // if (id == 1234) new Throwable().printStackTrace()
-    
+  def uniqueId_=(id: Int): Unit = {
+    if (Positioned.debugId == id) {
+      def printTrace() = {
+        val stack = Thread.currentThread().getStackTrace().map(">   " + _)
+        System.err.println(stack.mkString(s"> Debug tree (id=${Positioned.debugId}) creation \n> $this\n", "\n", "\n"))
+      }
+      printTrace()
```
Comment (anatoliykmetyuk) [inline compiler/src/dotty/tools/dotc/ast/Positioned.scala]: Why is the logic abstracted into a method instead of inlining it?

**Comment by nicolasstucki** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/ast/Positioned.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/ast/Positioned.scala]: `uniqueId_=` is probably a hot path as each tree creation will have a call to it. I put the code contained in this debug branch to make the bytecode of `uniqueId_=` smaller and probably simple to optimize from the JVM perspective.

**Pull request merged**